services/pwshow/show.go

```diff
@@ -50,22 +60,166 @@ func AddToList(u []string, p []string, c []string, user string) UserList {
 			log.Fatal(err)
 		}
 		U.Username, U.Password, U.Category = u[x], pwencrypter.Decrypt(dst[:n], decKey), c[x]
+		id, _ := strconv.Atoi(i[x])
+		cU.ID, cU.Creds.Username, cU.Creds.Password, cU.Creds.Category = id, u[x], pwencrypter.Decrypt(dst[:n], decKey), c[x]
 		FinalList = append(FinalList, U)
+		compareList = append(compareList, cU)
 	}
 	return FinalList
 }
 
+//GetPWID get password ID
+func GetPWID(user, pw, catg string) int {
+	if len(compareList) != 0 {
+		for _, n := range compareList {
+			if n.Creds.Username == user && n.Creds.Password == pw && n.Creds.Category == catg {
+				return n.ID
+			}
+		}
+	}
+	return 0
+}
+
 // ShowCreds Function Get user Stuff From the database
 func ShowCreds(user string, category string) UserList {
 	var Fl UserList
 	db := sqlite.InitDb()
 	uid := sqlite.GetUID(user, db)
 	if category != "" {
-		_, u, p, c := sqlite.GetStuff(uid, category, db)
-		Fl = AddToList(u, p, c, user)
+		i, u, p, c := sqlite.GetStuff(uid, category, db)
+		Fl = AddToList(i, u, p, c, user)
 	} else {
-		_, u, p, c := sqlite.GetStuff(uid, category, db)
-		Fl = AddToList(u, p, c, user)
+		i, u, p, c := sqlite.GetStuff(uid, category, db)
+		Fl = AddToList(i, u, p, c, user)
 	}
 	return Fl
 }
+
+//Compare compare front-end data and  back-end's
+func Compare(Flist UserList, category string) (map[UserStuff]UserStuff, UserList, UserList, bool) {
+	m := make(map[UserStuff]UserStuff)
+	var save UserList
+	var delt UserList
+	var ctg bool
+	/* if length of the data in front-end equal to the data in backend */
+	if len(FinalList) == len(Flist) {
+		mm, _, isctg := oldVsNew(FinalList, Flist, category, 0)
+		m = mm
+		ctg = isctg
+
+		/* if length of front-end data is bigger than the data in the backend */
+	} else if len(FinalList) < len(Flist) {
+		Fl := Flist[:len(FinalList)]
+		mm, _, isctg := oldVsNew(FinalList, Fl, category, 0)
+		if mm != nil {
+			m = mm
+			ctg = isctg
+		}
+		_, add, isctg := oldVsNew(FinalList, Flist, category, 1)
+		if add != nil {
+			save = add
+			ctg = isctg
+		}
+		/* if length of front-end's data is less than the backend's */
+	} else if len(FinalList) > len(Flist) {
+		Fl := make(UserList, len(FinalList))
+		Fl = clone(Flist, Fl)
+		_, deleted, isctg := oldVsNew(FinalList, Fl, category, 2)
+		delt = deleted
+		ctg = isctg
+	}
+	return m, save, delt, ctg
+}
+
+func clone(arr1, arr2 UserList) UserList {
+	for i := range arr1 {
+		arr2[i] = arr1[i]
+	}
+	return arr2
+}
+
+func oldVsNew(old, New UserList, category string, mode int) (map[UserStuff]UserStuff, UserList, bool) {
```

Comment:
Could you please add a commentary on what `oldVsNew` is doing.

I think you should also define a `type` for `mode` and use constants for the code to be clearer.

I feel like the logic of this function may break with new features. Though, I didn't managed to understand what it actually does yet!

Reply:
Yea i agree with you this function need a little bit of cleaning .. i let it like that because it does what i need it for which is comparing data comming from table in the front-end with our database data and give as an output ```map[UserStuff]UserStuff ``` this map give information about the values to be updated in our DB , for the ```UserList``` it holds additional values that we should save into the database, & finally for the last ```UserList``` is the values that we should delelte from the database .